README.md
[![Build Status](https://travis-ci.com/SmartcitySantiagoChile/dataUploader.svg?branch=master)](https://travis-ci.com/SmartcitySantiagoChile/dataUploader)   [![Coverage Status](https://coveralls.io/repos/github/SmartcitySantiagoChile/dataUploader/badge.svg?branch=master)](https://coveralls.io/github/SmartcitySantiagoChile/dataUploader?branch=master)

# datauploader

Python script to upload files to Elasticsearch

To run the uploader, first download the files in the
repo https://github.com/SmartcitySantiagoChile/elasticsearchInstaller/ and follow the instructions provided in the
Readme.

After elasticsearch and cerebro are running, go to the dataUploader folder and execute:

```shell
pip install -r requirements.txt
```

To load a file, the usage is:

```shell
python3 datauploader/loadData.py path/to/file
```

## Arguments:

    file: path to the file that contains the data to load.

## Optional:

    --host: the IP where ES is running. Default:"127.0.0.1".
    --port: the port that application is listening on. Default: 9200.
    --index: name of the index to create/use. If an index with the same name does not exist, the application will create it. Otherwise, it will use the existing index. By default, the name of the index is the extension of the file being uploaded.
    --chunk: number of docs to send to ES in one chunk. Default: 5000.
    --threads: number of threads to use. Default: 4.
    --timeout: timeout parameter of the ES client. Default:30 (in seconds.)

The default chunk size and number of threads are the ones that gave the best results when experimenting with different
files, so using the defaults is recommended. Nevertheless, sometimes loading a big file can cause a timeout error; in
this case, raising the timeout value should solve the issue (https://github.com/elastic/elasticsearch-py/issues/231)

## About the mappings:

The mappings must be in a folder called ```mappings``` in the same directory where ```loadData.py``` is. The script uses
the datafile extension to know what mapping to use so, if the mapping is called ```ext-template.json```, then the file
must have extension ```ext```. For the mappings, only JSON format is accepted.

As the ES team is planning to deprecate doctypes, the mappings used to load the file should all use the default doctype
that is ```doc```. If a property is a ```date```, then expliciting the format is recommended since ES is not able to
parse all date formats, and it may cause an exception. E.g.:

```json
{
  "mappings": {
    "doc": {
      "properties": {
        "id": {
          "type": "long"
        },
        "dateTime": {
          "type": "date",
          "format": "yyyy-MM-dd HH:mm:ss"
        },
        "type": {
          "type": "text"
        },
        "description": {
          "type": "text"
        }
      }
    }
  }
}
```

Additionally, is a property is listed as boolean only ```true``` and ```false``` values are accepted, using 0 and 1 will
cause a ```MapperParsingException```. In some cases when the data is not properly parsed the
option ```ignore malformed``` can be
used (https://www.elastic.co/guide/en/elasticsearch/reference/current/ignore-malformed.html,
check ```stop-template.json```.)

## About the extensions:

The extensions that this script loads are: ```expedition```, ```general```, ```od```, ```profile```, ```shape```
, ```speed```, ```stop``` and ```trip```.

In order to load a file that has a different extension its mapping must be included in the mappings folder and its
header must be added to ```datafile.py```. If not every column of the file is going to be loaded, if more columns need
to be added or if it's a nested file (meaning it has more than a line per document, like ```shape```), then a Python
file that inherits from ```datafile.py``` must be created and the ```getHeader``` and ```makeDocs``` methods must be
overwritten.

## Run tests:

Before to run tests you need to install test dependencies with the command:

```shell
pip install -r requirements-dev.txt
```

After that, you can execute:

```shell
python -m unittest
```

### run docker-compose tests

This test the process to upload to elasticsearch.

Build command:

```shell
docker-compose -p datauploader -f docker\docker-compose.yml build
```

Run command:

```shell
docker-compose -p datauploader -f docker\docker-compose.yml up --abort-on-container-exit
```

## Create a release version:

To create a release of this project, you must follow the next steps:

- Add a new version to `CHANGELOG.md` file with the following format:

```
vX.Y.Z YYYY-mm-DD

- Comment 1
- Comment 2
...
```

The most recent versions go at the top of the file.

- Change version code in `setup.py` with the same version added in `CHANGELOG.md`
- Commit changes of `CHANGELOG.md` file and `setup.py` file
- Create a tag with the same version added in `CHANGELOG.md`. Use the command `git tag vX.Y.Z`, avoid adding comments
  because it will be ignored it
- Push tag to GitHub with the command: `git push origin vX.Y.Z`

After these steps GitHub Actions take control, it will create a release.


CHANGELOG.md
# CHANGELOG

## v2.0.33 2023-07-27

- Increase timeout to 5 minutes
- Fix opdata loader
- Fix filter to upload trip data

## v2.0.32 2023-04-04

- Remove unused attribute in trip index

## v2.0.31 2023-04-04

- Apply filter only for dates greater than 2022-09-31

## v2.0.30 2023-04-03

- Cast tripNumber field to int for odbyroute index
- Ignore trip document if speed can not be calculated
- tviaje value for old files will be transformed to seconds
- Add dependencies to `setup.py` module

loadData.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import argparse
import glob
import os
import sys

from elasticsearch import Elasticsearch
from elasticsearch_dsl import Index

# add path so we can use function through command line
new_path = os.path.join(os.path.dirname(__file__), '..')
sys.path.append(new_path)

from errors import UnrecognizedFileExtensionError, IndexNotEmptyError
from uploader.bip import BipFile
from uploader.paymentfactor import PaymentFactorFile
from uploader.expedition import ExpeditionFile
from uploader.general import GeneralFile
from uploader.odbyroute import OdByRouteFile
from uploader.profile import ProfileFile
from uploader.shape import ShapeFile
from uploader.speed import SpeedFile
from uploader.stop import StopFile
from uploader.stopbyroute import StopByRouteFile
from uploader.trip import TripFile
from uploader.opdata import OPDataFile


def upload_file(es_instance, datafile, index_name=None, chunk_size=5000, threads=4, timeout=30):
    """ upload file to elasticsearch """

    # Get file extension
    index_name = os.path.basename(datafile).split(".")[1] if index_name is None else index_name

    # Determine file type according to the extension
    if index_name == 'expedition':
        uploader = ExpeditionFile(datafile)
    elif index_name == 'general':
        uploader = GeneralFile(datafile)
    elif index_name == 'odbyroute':
        uploader = OdByRouteFile(datafile)
    elif index_name == 'profile':
        uploader = ProfileFile(datafile)
    elif index_name == 'shape':
        uploader = ShapeFile(datafile)
    elif index_name == 'speed':
        uploader = SpeedFile(datafile)
    elif index_name == 'stop':
        uploader = StopFile(datafile)
    elif index_name == 'stopbyroute':
        uploader = StopByRouteFile(datafile)
    elif index_name == 'trip':
        uploader = TripFile(datafile)
    elif index_name == 'paymentfactor':
        uploader = PaymentFactorFile(datafile)
    elif index_name == 'bip':
        uploader = BipFile(datafile)
    elif index_name == 'opdata':
        uploader = OPDataFile(datafile)
    else:
        raise UnrecognizedFileExtensionError(datafile)

    # Load file to elasticsearch
    uploader.load(es_instance, index_name, chunk_size, threads, timeout)


def main():
    """
    This script will create an index on a existing elasticsearch instance and populate it using data from a given
    file. If an index with the same name already exists, it will use that index instead.
    """

    # Arguments and description
    parser = argparse.ArgumentParser(description='Add documents from a file to an elasticsearch index.')

    parser.add_argument('file', nargs='*', help='data file path, e.g. /path/to/file')
    parser.add_argument('--host', default="127.0.0.1", help='elasticsearch host, default is "127.0.0.1"')
    parser.add_argument('--port', default=9200, help='port, default is 9200')
    parser.add_argument('--index', help='name of the index to create/use')
    parser.add_argument('--chunk', default=5000, type=int, help='number of docs to send in one chunk, default is 5000')
    parser.add_argument('--threads', default=4, type=int, help='number of threads to use, default is 4')
    parser.add_argument('--timeout', default=30, type=int,
                        help='explicit timeout for each call, default is 30 (seconds)')
    args = parser.parse_args()

    # Get a client
    es = Elasticsearch(hosts=[{"host": args.host, "port": args.port}])

    # Give names to arguments
    index_name = args.index
    datafiles = args.file
    chunk_size = args.chunk
    threads = args.threads
    timeout = args.timeout

    # disable refresh
    if index_name is not None and es.indices.exists(index=index_name):
        Index(index_name, using=es).put_settings(body={'index.refresh_interval': -1})

    for datafile in datafiles:
        matched_files = glob.glob(datafile)
        for matched_file in matched_files:
            print('uploading file {0}'.format(matched_file))
            try:
                upload_file(es, matched_file, index_name, chunk_size, threads, timeout)
            except IndexNotEmptyError as e:
                # ignore it and continue uploading files
                print('Error: {0}'.format(e))

    # enable refresh
    if index_name is not None:
        Index(index_name, using=es).put_settings(body={'index.refresh_interval': '1s'})


if __name__ == "__main__":
    main()
